src/ROOTFrameReader.cc

```diff
@@ -0,0 +1,275 @@
+      for (size_t i = 0; i < refCollections->size(); ++i) {
+        const auto brName = root_utils::refBranch(name, i);
+        branches.refs[i] = root_utils::getBranch(catInfo.chain.get(), brName.c_str());
+      }
+    }
+
+    // vector members
+    if (auto* vecMembers = collBuffers.vectorMembers) {
+      for (size_t i = 0; i < vecMembers->size(); ++i) {
+        const auto brName = root_utils::vecBranch(name, i);
+        branches.vecs[i] = root_utils::getBranch(catInfo.chain.get(), brName.c_str());
+      }
+    }
+  }
+
+  // set the addresses and read the data
+  root_utils::setCollectionAddresses(collBuffers, branches);
+  root_utils::readBranchesData(branches, localEntry);
+
+  collBuffers.recast(collBuffers);
+
+  return collBuffers;
+}
+
+ROOTFrameReader::CategoryInfo& ROOTFrameReader::getCategoryInfo(const std::string& category) {
+  if (auto it = m_categories.find(category); it != m_categories.end()) {
+    // Use the id table as proxy to check whether this category has been
+    // initialized alrready
+    if (it->second.table == nullptr) {
+      initCategory(it->second, category);
+    }
+    return it->second;
+  }
+
+  // Use a nullptr TChain to signify an invalid category request
+  // TODO: Warn / log
+  static auto invalidCategory = CategoryInfo{nullptr};
+
+  return invalidCategory;
+}
+
+void ROOTFrameReader::initCategory(CategoryInfo& catInfo, const std::string& category) {
+  catInfo.table = std::make_shared<podio::CollectionIDTable>();
+  auto* table = catInfo.table.get();
+  auto* tableBranch = root_utils::getBranch(m_metaChain.get(), root_utils::idTableName(category));
+  tableBranch->SetAddress(&table);
+  tableBranch->GetEntry(0);
+
+  auto* collInfoBranch = root_utils::getBranch(m_metaChain.get(), root_utils::collInfoName(category));
+  auto collInfo = new std::vector<root_utils::CollectionInfoT>();
+  collInfoBranch->SetAddress(&collInfo);
+  collInfoBranch->GetEntry(0);
+
+  std::tie(catInfo.branches, catInfo.storedClasses) =
+      createCollectionBranches(catInfo.chain.get(), *catInfo.table, *collInfo);
+
+  delete collInfo;
+
+  // Finaly set up the branches for the paramters
+  root_utils::CollectionBranches paramBranches{};
+  paramBranches.data = root_utils::getBranch(catInfo.chain.get(), root_utils::paramBranchName);
+  catInfo.branches.push_back(paramBranches);
+}
+
+std::vector<std::string> getAvailableCategories(TChain* metaChain) {
+  auto* branches = metaChain->GetListOfBranches();
+  std::vector<std::string> brNames;
+  brNames.reserve(branches->GetEntries());
+  for (int i = 0; i < branches->GetEntries(); ++i) {
+    const std::string name = branches->At(i)->GetName();
+    const auto fUnder = name.find("___");
+    if (fUnder != std::string::npos) {
+      brNames.emplace_back(name.substr(0, fUnder));
+    }
+  }
+
+  std::sort(brNames.begin(), brNames.end());
+  brNames.erase(std::unique(brNames.begin(), brNames.end()), brNames.end());
+
+  return brNames;
+}
+
```

Comment:
```suggestion

std::map<std::string, detail::CollectionInfo> ROOTFrameReader::storedClasses(const std::string& catName) {
  std::map<std::string, detail::CollectionInfo> colls_info;
  if (auto& catInfo = getCategoryInfo(catName); catInfo.table) {
    for (const auto& [col, col_info] : catInfo.storedClasses)
      colls_info[col] = col_info;
  }
  return colls_info;
}

```

Reply:
@tmadlener Very nice work around these frames! May I suggest this addition to ease their integration into consuming workflows? This allows the access to the `TClass` objects as retrieved from the `TChain` on readout, and thus allows retrieving a human- (and `RDataSource`-) readable version of each collection type.

Reply:
Hi @forthommel,

The `detail::CollectionInfo` are currently really an implementation detail of the `ROOTFrameReader` and I am not entirely sure if they should be exposed in an interface in their current form. Definitely not while they are still in the `detail` namespace. What is the information that is actually necessary for `RDataSource` to work nicely here? I would prefer a solution that only exposes what is really necessary instead of spilling everything here.

Reply:
Hi @tmadlener, and thanks for the feedback!
What my scheme would need is the collection/category type with name `catName`, as stored in the `TChain`. Unless it is already exposed elsewhere in the API and I missed it?

Reply:
There is nothing in the API yet that exposes any of those details, as the currently foreseen EDM API does not need anything like this. Additionally, it gives podio the freedom to treat I/O implementations as a detail instead of making e.g. the Branch layout in the root files part of the API.

IIUC, then you would need the XYZData (resp. vectors of them), right?

I would suggest to move that discussion/implementation to another PR entirely, because this one is (too) large already and we would actually like to keep this one in its current form to make it possible to implement Schema evolution on top of this before iterating further.